crates/cli/src/cli/keygen.rs

```diff
@@ -1,48 +1,106 @@
             );
-            print!("Passphrase: ");
-            io::stdout().flush().expect("Could not flush stdout");
-        }
-        let passphrase1 = rpassword::read_password().unwrap();
-        if !quiet {
-            print!("Re-enter passphrase: ");
             io::stdout().flush().expect("Could not flush stdout");
+            get_secure_string_double_check("keystore Passphrase", quiet)
+                .expect("Could not retrieve passphrase")
         }
-        let passphrase2 = rpassword::read_password().unwrap();
-        if passphrase1 != passphrase2 {
-            println!("Passphrases do not match. Please retry...");
-            ::std::process::exit(1);
-        }
-        passphrase1
-    });
+    };
 
-    if !quiet {
-        println!("Generating keystore (this will take a few moments)...");
-    }
-    let (keystore, pub_key) = Keystore::new_standalone(mock_passphrase_manager(passphrase), None)?;
+    let (keystore, pub_key) = if let Some(derivation_index) = device_derivation_index {
+        user_prompt("This keystore is to be generated from a DPKI root seed. You can regenerate this keystore at any time by using the same root key mnemonic and device derivation index.", quiet);
+
+        let root_seed_mnemonic = root_seed_mnemonic.unwrap_or_else(|| {
+            get_secure_string_double_check("Root Seed Mnemonic", quiet)
+                .expect("Could not retrieve mnemonic")
+        });
+
+        match root_seed_mnemonic.split(' ').count() {
+            24 => {
+                // unencrypted mnemonic
+                user_prompt(
+                    "Generating keystore (this will take a few moments)...",
+                    quiet,
+                );
+                keygen_dpki(
+                    root_seed_mnemonic,
+                    None,
+                    derivation_index,
+                    keystore_passphrase,
+                )?
+            }
+            48 => {
+                // encrypted mnemonic
+                let mnemonic_passphrase = mnemonic_passphrase.unwrap_or_else(|| {
+                    get_secure_string_double_check("Root Seed Mnemonic passphrase", quiet)
+                        .expect("Could not retrieve mnemonic passphrase")
+                });
+                user_prompt(
+                    "Generating keystore (this will take a few moments)...",
+                    quiet,
+                );
+                keygen_dpki(
+                    root_seed_mnemonic,
+                    Some(mnemonic_passphrase),
+                    derivation_index,
+                    keystore_passphrase,
+                )?
+            }
+            _ => panic!(
+                "Invalid number of words in mnemonic. Must be 24 (unencrypted) or 48 (encrypted)"
+            ),
+        }
+    } else {
+        user_prompt(
+            "Generating keystore (this will take a few moments)...",
+            quiet,
+        );
+        keygen_standalone(keystore_passphrase)?
+    };
```

Comment:
Noticed a few panics in here even though we are returning an HCResult, is this because we cannot recover from them if they happen?

Reply:
Agreed, these expects should be replaced with `?`